### ISSUE TITLE
Container Timing API

### ISSUE THREAD

**Issue by keithamus** [w3ctag/design-reviews#1034]
こんにちは TAG-さん!

I'm requesting an early TAG design review of the Container Timing API.

The Container Timing API enables monitoring when annotated sections of the DOM are displayed on screen and have _finished their initial paint_.

  - Explainer¹:https://github.com/bloomberg/container-timing
  - User research:
  - Security and Privacy self-review²: This is essentially an extension of the Element Timing API (https://github.com/w3ctag/design-reviews/issues/326) so the same security constraints apply
  - GitHub repo:https://github.com/bloomberg/container-timing
  - Primary contacts:
      - Jason Williams (@jasonwilliams), Bloomberg, Author
      - José Dapena Paz (@jdapena), Igalia, Author
  - Organization/project driving the design: Bloomberg
  - Multi-stakeholder feedback³:
    - Chromium comments:
    - Mozilla comments: https://github.com/mozilla/standards-positions/issues/1155
    - WebKit comments: https://github.com/WebKit/standards-positions/issues/442
    - positive signals from developers (GitHub is supportive)

Further details:

  - [x] I have reviewed the TAG's [Web Platform Design Principles](https://www.w3.org/TR/design-principles/)
  - The group where the incubation/design work on this is being done (or is intended to be done in the future): This will likely move to WICG or the Web Perf WG.
  - The group where standardization of this work is intended to be done ("unknown" if not known): This will likely move to WICG or the Web Perf WG.
  - Existing major pieces of multi-implementer review or discussion of this design:
  - Major unresolved issues with or opposition to this design: 
  - This work is being funded by: Bloomberg

You should also know that... y'all are great, keep up the great work!

**Comment by torgo** [comment]
Thanks for bringing this our way. Regarding privacy, security - you've noted that "this is essentially an extension of the Element Timing API". That's understood. Since that was reviewed in 2018, have there been any additional issues raised with element timing API regarding privacy or security that might inform the design of this extension?

**Comment by jasonwilliams** [comment]
Hi @torgo, thanks for taking a look at this.

> have there been any additional issues raised with element timing API regarding privacy or security that might inform the design of this extension?

Yes, there is the [`PaintTimingMixin`](https://github.com/w3c/paint-timing/issues/62) which was added to `PerformanceElementTiming ` last year and also  [`exposed`](https://github.com/w3c/largest-contentful-paint/issues/91) to `largest-contentful-paint`. This does expose some extra information around paint timing to both and its unsure at this time whether Container Timing would need these extra properties right now. There have been several learnings were recently applied to element timing around security, and container timing should be following all of them.

Secondly, Container Timing doesn't expose any new information on top of element timing, it just aggregates the same information for a container. All of the properties listed [here](https://github.com/bloomberg/container-timing?tab=readme-ov-file#performancecontainertiming) can be achieved by using element timing on all elements within the same area.

@mmocny and @jdapena should have more information if needed

**Comment by jdapena** [comment]
About the Mixin:
- PerformanceContainerTiming entries should support the PaintTimingMixin. Internally we are already storing it, and that would be exposed the same way. Main thing is that we need to also provide a way to get the paint and presentation times for the first paint.
- Right now PerformanceContainerTiming.startTime is the last presentation time and lastPaintedElement is an element that was presented at that time. We need to decice if we want to have lastPresentedElement and lastPaintedElement. And we need to specify that startTime is the last presentation time explicitely.
- About firstRenderTime, maybe it should also be firstPresentationTime and firstPaintTime, and drop firstRenderTime. This way we will be able to get a complete view of presentation and painting.

Now, about the security/privacy: we need to incorporate the coarsening algorithm in the specification. I also need to review the native implementation (we are already using as input the coarsened numbers used for ElementTiming, so we should be already fine).

**Comment by jdapena** [comment]
> Hi all. We are looking at this in our TAG breakout. Two quick questions (and then likely more to follow):
> 
> 1. Do you think this adds to the fingerprinting surface area for a user (or UA context)?

As this can be polyfilled on top of Element Timing, it does not really add any new fingerprinting surface. Same considerations should be made for intersection rect and timers. The native implementations should be using coarsened timers as in ElementTiming.

> 3. How do you expect the coarsening algorithm to work here? Can you spell that out for us?

I think the Element Timing API security and privacy section applies pretty well here: https://w3c.github.io/element-timing/#sec-security

We will need to add a reference to it or repeat most of the Element Timing wording for the Container Timing render time details.

**Comment by torgo** [comment]
Hi folks - We discussed today and agreed to close this. We're happy to see this move forward. Thanks for being responsive to our concerns.
